### PR TITLE
add resource polardb endpoint

### PR DESCRIPTION
**Pull request by jokerhenson**
图片

### CODE REVIEW

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+				Optional: true,
+				Computed: true,
+			},
+			"read_write_mode": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{"ReadWrite", "ReadOnly"}, false),
+				Optional:     true,
+				Computed:     true,
+			},
+			"auto_add_new_nodes": {
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:48]: add default value please

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+	endpointType := d.Get("endpoint_type").(string)
+
+	request := polardb.CreateCreateDBClusterEndpointRequest()
+	request.RegionId = client.RegionId
+	request.DBClusterId = clusterId
+	request.EndpointType = endpointType
+
+	enpoints, err := polarDBService.DescribePolarDBInstanceNetInfo(clusterId)
+	if err != nil {
+		return nil
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:76]: handle the error please

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+			return resource.NonRetryableError(err)
+		}
+		addDebug(request.GetActionName(), raw, request.RpcRequest, request)
+		return nil
+	})
+
+	if err != nil {
+		return WrapErrorf(
+			err, DefaultErrorMsg, "alicloud_polardb_endpoint", request.GetActionName(), AlibabaCloudSdkGoERROR)
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:102]: format in one line

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	}
+	for _, item := range object {
+		if item.DBEndpointId == dbEndpointId {
+			d.Set("db_cluster_id", dbClusterId)
+			d.Set("endpoint_type", item.EndpointType)
+			d.Set("db_nodes", item.Nodes)
+			d.Set("auto_add_new_nodes", item.AutoAddNewNodes)
+			d.Set("read_write_mode", item.ReadWriteMode)
+			endpointConfig := make(map[string]string)
+			err = json.Unmarshal([]byte(item.EndpointConfig), &endpointConfig)
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:159]: handle the error please

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+		addDebug(request.GetActionName(), raw, request.RpcRequest, request)
+		return nil
+	})
+
+	if err != nil {
+		return WrapErrorf(
+			err, DefaultErrorMsg, "alicloud_polardb_endpoint", request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
+	deadline := time.Now().Add(time.Duration(DefaultTimeoutMedium) * time.Second)
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:105]: WaitForPolarDBEndpoints

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+		}
+	}
+
+	return nil
+}
+
+func resourceAlicloudPolarDBEndpointUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	polarDBService := PolarDBService{client}
+	d.Partial(true)
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:170]: no need Partial here

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	})
+
+	if err != nil {
+		if IsExpectedErrors(err, []string{"InvalidDBClusterId.NotFound", "InvalidCurrentConnectionString.NotFound", "AtLeastOneNetTypeExists"}) {
+			return nil
+		}
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
+	return nil
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:264]: add waitfor method call please

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{"Custom"}, false),
+				ForceNew:     true,
+			},
+			"db_nodes": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"read_write_mode": {
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:42]: add default value please

File: website/docs/r/polardb_endpoint.html.markdown [hunk 9]
```diff
@@ -0,0 +1,82 @@
+---
+subcategory: "PolarDB"
+layout: "alicloud"
+page_title: "Alicloud: alicloud_polardb_endpoints"
```
Comment (Lexsss) [inline website/docs/r/polardb_endpoint.html.markdown:4]: alicloud_polardb_endpoint

File: website/docs/r/polardb_endpoint.html.markdown [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+sidebar_current: "docs-alicloud-resource-poalrdb-endpoint"
+description: |-
+  Provides a PolarDB instance endpoint resource.
+---
+
+# alicloud\_polardb\_endpoint
+
+Provides a PolarDB endpoint resource to allocate an Internet endpoint string for PolarDB instance.
+
+-> **NOTE:** Available in v1.77.0+. Each PolarDB instance will allocate a intranet connection string automatically and its prefix is Cluster ID.
```
Comment (Lexsss) [inline website/docs/r/polardb_endpoint.html.markdown:14]: 1.79.0

File: website/docs/r/polardb_endpoint.html.markdown [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+## Argument Reference
+
+The following arguments are supported:
+
+* `db_cluster_id` - (Required, ForceNew) The Id of cluster that can run database.
+* `endpoint_type` - (Required, ForceNew) Type of endpoint. Valid value: `Custom`. Currently supported only `Custom`.
+* `read_write_mode` - (Optional) Read or write mode. Valid values are `ReadWrite`, `ReadOnly`. Default to `ReadOnly`.
+* `db_nodes` - (Optional) Node list for connection address configuration, string type, separate node ids with (,). At least 2 nodes if specified, or if the cluster has more than 3 nodes, read-only endpoint is allowed to mount only one node. Default is all nodes.
+* `auto_add_new_nodes` - (Optional) Whether the new node automatically joins the default cluster address. Valid values are `Enable`, `Disable`. Default to `Disable`.
+* `endpoint_config` - (Optional, MapString) Advanced configuration of the cluster address.
```
Comment (Lexsss) [inline website/docs/r/polardb_endpoint.html.markdown:68]: remove the `MapString `

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+		configItem["AutoAddNewNodes"] = d.Get("auto_add_new_nodes").(string)
+		update = true
+	}
+	if d.HasChange("endpoint_config") {
+		endpointConfig, _ := json.Marshal(d.Get("endpoint_config"))
+		request.EndpointConfig = string(endpointConfig)
+		configItem["EndpointConfig"] = string(endpointConfig)
+		update = true
+	}
+	if update {
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint.go:186]: The update flag is redundant

File: alicloud/resource_alicloud_polardb_endpoint_test.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+
+		// module name
+		IDRefreshName: resourceId,
+
+		Providers:    testAccProviders,
+		CheckDestroy: rac.checkResourceDestroy(),
+		Steps: []resource.TestStep{
```
Comment (Lexsss) [inline alicloud/resource_alicloud_polardb_endpoint_test.go:41]: Modify all modifiable fields

File: alicloud/service_alicloud_polardb.go [hunk 14]
```diff
@@ -317,6 +317,141 @@ func (s *PolarDBService) WaitForPolarDBConnection(id string, status Status, time
...
+	newEndpoint["endpoint_type"] = endpointType
+	newEndpoint["read_write_mode"] = readWriteMode
+	newEndpoint["auto_add_new_nodes"] = autoAddNewNodes
+	newEndpoint["nodes"] = nodes
+
+	deadline := time.Now().Add(time.Duration(timeout) * time.Second)
+	for {
+		endpoints, err := s.DescribePolarDBInstanceNetInfo(dbClusterId)
+		if err != nil {
+			return nil, err
```
Comment (Lexsss) [inline alicloud/service_alicloud_polardb.go:426]: WrapError(err)

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+
+	enpoints, err := polarDBService.DescribePolarDBInstanceNetInfo(clusterId)
+	if err != nil {
+		return WrapError(err)
+	}
+	oldEndpoints := make([]interface{}, 0)
+	for _, value := range enpoints {
+		oldEndpoints = append(oldEndpoints, value.DBEndpointId)
+	}
+	oldEndpointIds := schema.NewSet(schema.HashString, oldEndpoints)
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_polardb_endpoint.go:81]: What does the `oldEndpointIds` do?

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+		return nil
+	})
+
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, "alicloud_polardb_endpoint", request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
+	dbEndpointId, err := polarDBService.WaitForPolarDBEndpoints(d, Active, oldEndpointIds, DefaultTimeoutMedium)
+	if err != nil {
+		d.SetId("")
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_polardb_endpoint.go:106]: Why setting the id to ""?

File: alicloud/resource_alicloud_polardb_endpoint.go [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+	if d.HasChange("read_write_mode") {
+		request.ReadWriteMode = d.Get("read_write_mode").(string)
+		configItem["ReadWriteMode"] = d.Get("read_write_mode").(string)
+	}
+	if d.HasChange("auto_add_new_nodes") {
+		request.AutoAddNewNodes = d.Get("auto_add_new_nodes").(string)
+		configItem["AutoAddNewNodes"] = d.Get("auto_add_new_nodes").(string)
+	}
+	if d.HasChange("endpoint_config") {
+		endpointConfig, _ := json.Marshal(d.Get("endpoint_config"))
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_polardb_endpoint.go:171]: Please don't hidden the error

File: alicloud/service_alicloud_polardb.go [hunk 18]
```diff
@@ -317,6 +317,118 @@ func (s *PolarDBService) WaitForPolarDBConnection(id string, status Status, time
...
+		if time.Now().After(deadline) {
+			return WrapErrorf(err, WaitTimeoutMsg, id, GetFunc(1), timeout, endpoint, item, ProviderERROR)
+		}
+		time.Sleep(DefaultIntervalShort * time.Second)
+	}
+
+	return nil
+}
+
+func (s *PolarDBService) WaitForPolarDBEndpoints(d *schema.ResourceData, status Status, endpointIds *schema.Set, timeout int) (*string, error) {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_polardb.go:385]: why return *string not string

File: alicloud/resource_alicloud_polardb_endpoint.go (hunk 15 above)
Comment (jokerhenson) [reply alicloud/resource_alicloud_polardb_endpoint.go:81]: Openapi doesn't return the newly created endpoint id, before creating, get the id of the existing endpoints to determine the newly created endpoint id.

File: alicloud/service_alicloud_polardb.go (hunk 18 above)
Comment (jokerhenson) [reply alicloud/service_alicloud_polardb.go]: Has been modified to string.

File: alicloud/resource_alicloud_polardb_endpoint.go (hunk 16 above)
Comment (jokerhenson) [reply alicloud/resource_alicloud_polardb_endpoint.go]: unnecessary to set the id to "", has been deleted.

**Comment by jokerhenson** [comment]
图片

**Pull request merged**